HylaFAX+: init at 5.5.8

RamKromberg opened the pull request:
###### Motivation for this change

trees?
###### Things done

Receiving and sending faxes works.
Email redirection works.
Optional hfaxd client-server service works (mostly for LAN clients) and some local GUIs.

Configuration is somewhat manual\interactive due to the nature of the hardware:
- 1. In _configuration.nix_: `services.hylafaxp.enable = true;`
- 2. As root: _faxsetup_.
- 3. Copy _/var/spool/hylafax/etc/config_ and _/var/spool/hylafax/etc/config.tty*_ to _configuration.nix_. e.g.

```
  services.hylafaxp = {
    enable = true;
    config = ''
      LogFacility:                daemon
      CountryCode:                011
      AreaCode:           01
      LongDistancePrefix: 1
      InternationalPrefix:        011
      DialStringRules:    etc/dialrules
      ServerTracing:              1
    '';
    devs = {
      ttyACM0 = ''
        CountryCode:        011
        AreaCode:       01
        FAXNumber:      12345678
        LongDistancePrefix: 1
        InternationalPrefix:    011
        DialStringRules:    etc/dialrules
        ServerTracing:      1
        SessionTracing:     0xFFF
        RecvFileMode:       0600
        LogFileMode:        0600
        DeviceMode:     0600
        RingsBeforeAnswer:  5
        SpeakerVolume:      off
        GettyArgs:      "-h %l dx_%s"
        LocalIdentifier:    companyOrHome
        TagLineFont:        etc/LiberationSans-25.pcf
        TagLineFormat:      "From %%l|%c|Page %%P of %%T"
        MaxRecvPages:       25
        #
        #
        # Modem-related stuff: should reflect modem command interface
        # and hardware connection/cabling (e.g. flow control).
        #
        ModemType:      Class1      # use this to supply a hint
        ModemRate:      19200       # rate for DCE-DTE communication
        ModemFlowControl:   rtscts      # default
        #
        ModemNoFlowCmd:     AT&K0       # setup no flow control
        ModemHardFlowCmd:   AT&K3       # setup hardware flow control
        ModemSoftFlowCmd:   AT&K4       # setup software flow control
        ModemSetupDTRCmd:   AT&D2       # setup so DTR drop resets modem
        ModemSetupDCDCmd:   AT&C1       # setup so DCD reflects carrier (or not)
        #
        ModemMfrQueryCmd:   !Rockwell
        ModemModelQueryCmd: !RC288DPi
        ModemRevQueryCmd:   ATI3        # product information
        #
        ModemDialCmd:       ATDT%s      # T for tone dialing
      '';
    };
  };
```
- 4. _nixos-rebuild switch_.

This will produce persistent results between reboot & updates.
- 5. (Optional) For email dispatching, it's enough to add:

```
    faxDispatch = ''
      SENDTO=myemail@mydomain.com;
      FILETYPE=pdf;
    '';
```

to services.hylafaxp and

```
  services.postfix = {
    enable = true;
    setSendmail = true;
  };
```
- 6. (Optional) For the client-server protocol service, adding

```
    hfaxd = {
      enable = true;
      hosts = ''
        localhost
        127.0.0.1
      '';
    };
```

to services.hylafaxp should be sufficient for local access. See faxadduser for LAN clients and passwords.

I've only tested one soft-modem, but I wrote the module to support multiple modems.

Built and tested with `nixos-rebuild switch -I nixpkgs=/home/user/nixpkgs`.
- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Note: HylaFAX vs. HylaFAX+ is 2012 vs. 2016.

RamKromberg committed:
hylafaxp: init at 5.5.8

Mic92 commented on nixos/modules/services/networking/hylafaxp.nix:
```diff
@@ -0,0 +1,187 @@
+    ${pkgs.ghostscript}/bin/gs -q -sDEVICE=tiffg3 -sFONTPATH="${pkgs.ghostscript}/share/ghostscript/fonts ${pkgs.ghostscript}/share/ghostscript/${pkgs.ghostscript.version}/Resource/Init ${pkgs.ghostscript}/share/ghostscript/${pkgs.ghostscript.version}/lib ${pkgs.ghostscript}/share/ghostscript/${pkgs.ghostscript.version}/Resource/Font ${pkgs.ghostscript}/share/ghostscript/fonts ${pkgs.ghostscript} ...
+    ${concatStringsSep ";" (zipListsWith (a: b: "cp " + a + " /var/spool/hylafax/etc/config." + b) etcConfigDevs etcDevs)};
+    chmod 0755 /var/spool/hylafax
+    ls -A /var/spool/hylafax/ | grep -v -E 'recvq|log' | ${pkgs.gawk}/bin/awk '{print "/var/spool/hylafax/"$1}' | xargs chmod -R 0755
```
something with `find` would be cleaner:

```
find /var/spool/hylafax -type d \( -path recvq -o -path log \) -prune -o -print0 | xargs -0 chmod -R 0755
```

RamKromberg replied:
@Mic92 It's also good but I prefer it my way since it's easier to understand and maintain.

@joachifm I've considered this previously and decided I'd like to avoid bash-only features when possible.

If you guys can produce a solution that doesn't depend on regex matching (compute) or that walks only the one time (disk io), I'm sticking with the choice I find easiest to maintain.

Mic92 commented on pkgs/servers/hylafaxp/default.nix:
```diff
@@ -0,0 +1,290 @@
+    {
+       ${coreutils}/bin/echo -n "$* "
+    }
+          ttyPort()
```
indentation?

RamKromberg replied:
@Mic92 it's the indentation faxsetup outputs. Same goes for hyla.conf, setup.cache and setup.modem.

I'd rather not touch them for diffing purposes if a future update changes them.

RamKromberg commented:
WIP: Taking a look at the email redirection.

Mic92 commented on nixos/modules/services/networking/hylafaxp.nix:
```diff
@@ -0,0 +1,262 @@
+    services.hylafaxp.pagesizes = mkOption {
+      type = types.lines;
+      default = ''
+        #	$Id: pagesizes.in 2 2005-11-11 21:32:03Z faxguy $
```
This file did not change since 2005. Do you think it is necessary to provide an option to change it?

RamKromberg replied:
Yeah it's the paper size you send.
HylaFAX usually distributes with Letter as a default but I've taken the liberty to use ISO A4 instead since that's what I, and presumably, most of the NixOS user base, uses.

Anyhow, Americans\Japanese will want to change this.

Mic92 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -6081,6 +6081,11 @@ in
 
+  hylafaxp = callPackage ../servers/hylafaxp {
+    pam = null;
+    openldap = null;
```
well, if those dependencies are never used in `hylafax`, could we not drop them?

RamKromberg replied:
Since ldap and pam aren't useful to me I nulled them out.

But, a more enterprise-centric person might want to look at the module for hfaxd's inetd support that's likely to go with pam. Similarly, ldap...

Mic92 commented on nixos/modules/services/networking/hylafaxp.nix:
```diff
@@ -0,0 +1,262 @@
+    ${if (config.services.hylafaxp.config != "") then "cat ${_config} - > /var/spool/hylafax/etc/config" else ""}
+    ${if (config.services.hylafaxp.hfaxd.hosts != "") then "cat ${_hosts_hfaxd} - > /var/spool/hylafax/etc/hosts.hfaxd" else ""}
+    ${if (config.services.hylafaxp.faxDispatch != "") then "cat ${_FaxDispatch} - > /var/spool/hylafax/etc/FaxDispatch" else ""}
+    ${pkgs.ghostscript}/bin/gs -q -sDEVICE=tiffg3 -sFONTPATH="${pkgs.ghostscript}/share/ghostscript/fonts ${pkgs.ghostscript}/share/ghostscript/${pkgs.ghostscript.version}/Resource/Init ${pkgs.ghostscript}/share/ghostscript/${pkgs.ghostscript.version}/lib ${pkgs.ghostscript}/share/ghostscript/${pkgs.ghostscript.version}/Resource/Font ${pkgs.ghostscript}/share/ghostscript/fonts ${pkgs.ghostscript} ...
```
Can those files above replaced by `writeText` and symlinks?

RamKromberg replied:
We need ghostscript to produce font names and paths depending on what's available on the system and the specific ghostscript package version.

RamKromberg commented on nixos/modules/services/networking/hylafaxp.nix:
```diff
@@ -0,0 +1,262 @@
+          assert dev != "";
+          nameValuePair "hylafaxp-faxgetty-${dev}" {
+            description = "faxgetty ${dev}: HylaFAX+ front-door process";
+            after = optionals (dev!="dummy") [ "hylafaxp-setup.service" "hylafaxp-faxq.service" "${dev}.device" ];
```
@Mic92 This one sucks but I'm not sure how to clean it up.

I'd like to condition this entire let..in.. construct by config.services.hylafaxp.devs and move the error message* to the setupscript.

It's quite ugly and convoluted in an already complex structure but I don't know how to do the condition so I ended up recycling the unit to display the error letting the user know they should run faxsetup and update config.services.hylafaxp.devs accordingly so the unit won't be a total waste.

Similarly, there are also a few setupScript config.services.hylafaxp.devs if statements I'm not particularly happy about.

RamKromberg commented:
@Mic92 I've cleaned up a few problem spots including the FONTPATH you've requested. It's still being produced by ghostscript since we still need to account for different fonts and changes between ghostscript versions... But it's nicer looking :)

@lethalman Thanks for explaining how to condition the let...in... statement.

Mic92 commented:
@RamKromberg 
I have refactored the module:
https://github.com/Mic92/nixpkgs/commit/babb76722c1355aa0d84ccd9847bc69842afb3fb

7c6f434c commented:
@Mic92 Have you ever pushed the refactored version to mainline `Nixpkgs`?

Mic92 commented:
@7c6f434c no

7c6f434c commented:
@Mic92 any reason not to push it (and close this PR afterwards)?

RamKromberg commented:
@7c6f434c If memory serves, it was suggested the package should probably be rewritten (and updated*) to run in a container since it handles its own ring0->ring3 de-escalation. I think I tried doing it myself but run into some issues... Well, I was pretty satisfied with the C code security-wise and was running it for a few months without any major problems so I just never got around to doing it.

Eventually, it stalled as I became "distracted" with unrelated low-level work and generally gone inactive with NixOS development.

Seeing how I moved my fax-modem to a RasPi and been running Debian there... Well, it would be a shame to see so much work going to waste but I'll understand if you'd want to close the pull request. Though I wish it could be archived or something in case someone ends up wanting to work on it or something.

* 5.5.9 http://hylafax.sourceforge.net/news/5.5.9.php

7c6f434c commented:
@RamKromberg I think the refactored version by @Mic92 addresses some of the discussion. I asked about closing the PR after merging whatever further work on top of this PR has been done.

In general, closed PRs are visible in a separate tab and are displayed in search; availability of the branch actually depends on whether you keep it in your fork.
